examples/aes-ccm/src/ccm_mode.rs

```diff
@@ -0,0 +1,197 @@
+        b[((k+tmp) as usize)-i] = U8(curr);
+        copy2 = copy2 >> 8;
+    }
+
+    k = k + tmp;
+
+    for i in 0..(alen as usize) {
+        let tmp2 = a.get_exact_chunk(1, i);
+        b = b.set_exact_chunk(1, i+(k as usize), &tmp2);
+    }
+
+    k = k + alen;
+
+    for _t in 0..16 {
+        if k % 16u64 != 0u64 {
+            // add zero padding for Associated Data
+            b[k as usize] = U8(0u8);
+            k = k + 1u64;
+        }
+    }
+
+    // creation of b(u+1) to b(r)
+    for i in 0..(plen as usize) {
+        let tmp2 = p.get_exact_chunk(1, i);
+        b = b.set_exact_chunk(1, i+(k as usize), &tmp2);
+    }
+
+    k = k + plen;
+
+    for _t in 0..16 {
+        if k % 16u64 != 0u64 {
+            // add zero padding for Payload
+            b[k as usize] = U8(0u8);
+            k = k + 1u64;
+        }
+    }
+
+    b
+}
+
+fn get_t(b: &ByteSeq, key: Key128, num: u8) -> ByteSeq {
+    let b0 = b.get_exact_chunk(16, 0);
+    let bloc = Block::from_seq(&b0);
+    let mut y_curr = aes128_encrypt_block(key, bloc);
+
+    for i in 1..b.len()/16 {
+        let mut b_curr = Block::from_seq(&b.get_exact_chunk(16, i));
+        b_curr = y_curr ^ b_curr;
+        y_curr = aes128_encrypt_block(key, b_curr);
+    }
+
+    ByteSeq::from_seq(&(y_curr.slice(0, num as usize)))
+}
+
+fn counter_func(n: &ByteSeq, nlen: u8, m: u64) -> ByteSeq {
+    let qlen: u8 = 15u8 - nlen;
+    let flag = qlen - 1u8;
+    let mut ctr = ByteSeq::new((16u64 * (m+1u64)) as usize);
+    let high = 255u64; // 0xFF
+
+    for i in 0..(m as usize)+1 {
+        let mut icopy = i as u64;
+
+        let k = 16 * i;
+        ctr[k] = U8(flag);
+
+        for j in 0..(nlen as usize) {
+            let tmp2 = n.get_exact_chunk(1, j);
+            ctr = ctr.set_exact_chunk(1, ((k+j) as usize)+1, &tmp2);
+        }
+
+        for x in 1..16-(nlen as usize)-1 {
+            let curr = icopy & high;
+            ctr[k+16-x] = U8(curr as u8);
+            icopy = icopy >> 8;
+        }
+    }
+
+    ctr
+}
+
+fn ctr_cipher(ctr: &ByteSeq, key: Key128, m: u64) -> (ByteSeq, ByteSeq) {
+    let ctr_zero = Block::from_seq(&ctr.get_exact_chunk(16, 0));
+    let s0 = ByteSeq::from_seq(&aes128_encrypt_block(key, ctr_zero));
+    let mut s = ByteSeq::new((16u64*m) as usize);
+
+    for i in 1..(m as usize)+1 {
+        let ctr_block = Block::from_seq(&ctr.get_exact_chunk(16, i));
+        let s_curr = aes128_encrypt_block(key, ctr_block);
+        let seq_s = ByteSeq::from_seq(&s_curr);
+        s = s.set_exact_chunk(16, i-1, &seq_s);
+    }
+
+    (s0, s)
+}
+
+pub fn encrypt_ccm(a: ByteSeq, n: ByteSeq, pay: ByteSeq, key: Key128, tlen: u8, alen: u64, nlen: u8, plen: u64) -> ByteSeq {
+    let b = format_func(&a, &n, &pay, tlen, alen, nlen, plen); // step 1
+    let t = get_t(&b, key, tlen); // steps 2 to 4
+
+    let m = (plen+15u64)/16u64; // round up
+    let counter = counter_func(&n, nlen, m);
+    let (s0, s) = ctr_cipher(&counter, key, m);
```

Comment:
Should I use a mutable reference? Or copy explicitly? If copy, then please share the syntax for that.

Reply:
It's working by passing `key` as a reference. I'll use it like that.

Reply:
mutable won't help. You don't want to mutate it. You need a copy or an immutable reference. Generally references are preferrable because you don't have to copy memory. But if you have to, just `clone` it.